File: topicos/dynamic-programming/group-1/binary_exponation/binary_exponation.ts
export function binaryExponation(base: bigint, exp: bigint): bigint {
    let result: bigint = 1n;

    while (exp > 0n) {
        if ((exp & 1n) === 1n) {
            result *= base;
        }

        base *= base;
        exp >>= 1n;
    }

    return result;
}

File: topicos/dynamic-programming/group-1/binary_exponation/binary_exponation.go
package main

import "fmt"

func binary_exponation(base int64, exp int) int64 {
	var result int64 = 1

	for exp > 0 {
		if exp & 1 == 1 {
			result *= base
		}

		base *= base
		exp >>= 1

	}

	return result
}